No jquery dependency on client-side code.

valdrinkoshi opened the pull request:
Fixes #631 
Remove jquery dependency from client side code.
This PR applies 2 fixes:
- replace jQuery APIs with native DOM APIs, some examples:
  - `jQuery(document).ready(...` => `document.addEventListener('DOMContentLoaded', ...`
  - `jQuery(element).on('click', ...` => `element.addEventListener('click', ...`
  - `jQuery(document).find(...` => `document.querySelectorAll(...`
- improve performance on `scroll/resize` events by debouncing callback to next animation frame

The admin client has been purposefully left untouched in this PR, since wordpress itself relies on jQuery.

The PR relies on `NodeList.forEach` to loop through elements, but that's not implemented in Internet Explorer, though a polyfill can be used in that case https://developer.mozilla.org/en-US/docs/Web/API/NodeList/forEach#Polyfill - let me know if that's needed.

valdrinkoshi commented on assets/js/homepage.js:
```diff
@@ -4,43 +4,29 @@
-	function homepageContentDimensions() {
-		var windowWidth	    = jQuery( window ).width();
-		var offset          = jQuery( '.site-main' ).offset();
+	requestAnimationFrame(function() {
```
Previously, the implementation would wait for the document to be ready. By using `requestAnimationFrame`, we  can assume that the first frame is about to be rendered, aka the parsing of the document is done.
This shouldn't be necessary, as the homepage script is injected at the end of the document (hence the DOM is already created and can be queried).

valdrinkoshi commented on assets/js/homepage.js:
```diff
@@ -62,22 +48,37 @@
-		homepageContentDimensions();
-	});
-} )();
+	function homepageContentDimensions() {
```
We delay the update of the styles to the next frame, as this method might be called very frequently during window resize events.

valdrinkoshi commented on assets/js/navigation.js:
```diff
@@ -6,27 +6,29 @@
-			jQuery( this ).parent().toggleClass( 'active' );
-			e.preventDefault();
+	// Add class to footer search when clicked.
+	requestAnimationFrame( function() {
```
Similar to homepage.js, this file is injected at the end of the document, so DOM is already available to be queried. We wait a RAF, but that shouldn't be needed.

valdrinkoshi commented on assets/js/navigation.js:
```diff
@@ -52,54 +54,30 @@
-			jQuery( this ).parents().toggleClass( 'focus' );
+	// Add focus class to li and cart dropdown
+	requestAnimationFrame(function() {
+		document.querySelectorAll('.main-navigation a, .secondary-navigation a, .site-header-cart a').forEach(function(anchor) {
```
I've merged the focus toggling in this method. Not clear what's the behavior we desire to have when on mobile, will need some assistance on this one.

valdrinkoshi commented on inc/woocommerce/class-storefront-woocommerce.php:
```diff
@@ -114,13 +114,16 @@ public function woocommerce_scripts() {
-					$( 'body' ).on( 'click', '#respond p.stars a', function() {
-						var $container = $( this ).closest( '.stars' );
-						$container.addClass( 'selected' );
+				requestAnimationFrame(function() {
```
Wasn't exactly sure where this script will be injected, so I've implemented it by assuming `p.stars` are already rendered by the next frame time (aka, no other stars elements are created dynamically at some later point)

claudiulodro commented:
This looks great. Everything works properly and I like the RAF improvements.

Couple notes:
1. It's going to be a little bit of a pain in the butt to go through and do this after everything's written, but [spacing in the JS needs to follow the WP style guidelines](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#examples-of-good-spacing). That's mostly going to be adding a space in function calls. For example, `(foo,bar) -> ( foo, bar )`

2. On Safari the checkout order review box sticks to the screen too early. This is probably because of a browser difference in how Safari calculates box heights or window heights. It's not a problem in other browsers (haven't tested Edge, though).
<img width="1422" alt="screen shot 2018-01-05 at 1 07 58 pm" src="https://user-images.githubusercontent.com/7317227/34628956-1c68ed00-f21b-11e7-9e88-b4e83ca50ed8.png">

@jameskoster @tiagonoronha How do you come up with the version number for the enqueued scripts (e.g. `'20120206'`)? These JS files will need a new version number.

tiagonoronha commented:
>On Safari the checkout order review box sticks to the screen too early. This is probably because of a browser difference in how Safari calculates box heights or window heights. It's not a problem in other browsers (haven't tested Edge, though).

We are removing that script very soon. See #637.

>How do you come up with the version number for the enqueued scripts (e.g. '20120206')? These JS files will need a new version number.

I don't think there's a particular logic behind that number. You can change that you use the version number of Storefront.

valdrinkoshi commented:
@claudiulodro thanks for the feedback. I will update the js space styling and give a look at the safari bug, just for fun 👌

valdrinkoshi commented:
@claudiulodro PTAL at the last 5 commits:
- fixed js space formatting
- now using storefront version for the js files
- fixed sticky positioning by simplifying the logic (it's enough to check for `getBoundingClientRect().top <= 0`, see last commit)

claudiulodro commented:
Excellent. Everything looks good and works well now. Thanks!
